language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,366 @@
+// tables. The order of tables can be anything.
+
+// Common section
+pub struct CompiledHeader {
+    // handles are references to definitions within and cross modules
+    pub module_handles: Vec<ModuleHandle>,
+    pub struct_handles: Vec<StructHandle>,
+    pub function_handles: Vec<FunctionHandle>,
+    pub field_handles: Vec<FieldHandle>,
+    // instantiations are references to a *generic* handle and and its instantiation, whether
+    // partial or complete
+    pub struct_instantiations: Vec<StructInstantiation>,
+    pub function_instantiations: Vec<FunctionInstantiation>,
+    pub field_instantiations: Vec<FieldInstantiation>,
+    // common blob tables
+    pub signatures: SignaturePool,
+    pub identifiers: IdentifierPool,
+    // constants
+    pub byte_array_pool: U8VectorPool,
+    pub address_pool: AddressPool,
+}
+
+// In `CompiledScript` we replace the FunctionDefinition for main(..) with broken up data
+// to set up the call frame and define/verify main signature and arguments
+pub struct CompiledScript {
+    pub header: CompiledHeader,
+
+    // Below should be enough information to materialize a
+    // main<Cup<T>, W>(arg1, ..., argn) {}
+    // with args of any type (including generic types) that the VM (verification
+    // and runtime) can use.
+    // There is a verification step here that is *purely* runtime.
+    // The client is passing the type arguments and you can imagine restrictions
+    // given our resource semantic.
+
+    // function arguments signature, main is void on return
+    pub arg_types: SignatureIndex,
+    // type parameter for the script
+    pub type_params: Vec<Kind>,
+    // code
+    pub main: CodeUnit,
+}
+
+// A CompiledModule is what it used to be but you will notice the removal of `FieldDefinition`s.
+// Fields are now serialized with the `StructDefinition` which changes verification around fields.
+// (see `StructDefinition` for details)
+pub struct CompiledModule {
+    pub header: CompiledHeader,
+
+    pub struct_defs: Vec<StructDefinition>,
+    pub function_defs: Vec<FunctionDefinition>,
+}
+
+//
+// Basics...
+//
+
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+// Any expressible type in Move
+pub enum Type {
+    Bool,
+    U8,
+    U64,
+    U128,
+    Address,
+    Vector(Box<Type>),
+    Struct(StructHandleIndex),
+    // a complete or partial instantiation of a generic type
+    StructInstantiation(StructInstantiationIndex),
+    Reference(Box<Type>),
+    MutableReference(Box<Type>),
+    // a type argument provided by the definition (via instantiation)
+    TypeParameter(TypeParameterIndex),
+}
+
+//
+// Handles
+//
+
+
+pub struct ModuleHandle {
+    pub address: AddressPoolIndex,
+    pub name: IdentifierIndex,
+}
+
+
+pub struct StructHandle {
+    pub module: ModuleHandleIndex,
+    pub name: IdentifierIndex,
+    pub is_nominal_resource: bool,
+    pub type_params: Vec<Kind>,
```

Comment:
this is `type_formals` in the current code base.
Should we rename it or leave` formals`?

Reply:
params or parameters (whatever we use above) seems good?